Building the MSUnmerged construction around functional pipelines and configurable RSE lists.

todor-ivanov opened the pull request:
Fixes #10599 

#### Status
ready

#### Description
In order to achieve the scalability goals we have put for the service we need to split the initial list of RSEs to work on by Pod. So we  decided to have the that as a per K8 pod configurable `rseExpression` which is supposed to produce the desired subset of RSEs when executed in a call to Rucio. For this `rseExpression` we have prepared a placeholder in the MSConfig, which is supposed to be substituted with the `rseExppression` provided by the .yaml file for the pod at run time right before service execution. 

This PRalso provides  the basic logic for running the load through a functional pipeline, while accumulating counters and providing summary at the end of the service polling cycle     

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
https://github.com/dmwm/CMSKubernetes/pull/612
https://github.com/dmwm/deployment/pull/1059/commits
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/89
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/90


#### External dependencies / deployment changes
We depend on the rest of the construction of the K8 pods and their setup.

todor-ivanov committed:
Building the MSUnmerged construction arround functional Plines and configurable RSE lists.

Remove documentSchema from docstring in MSUnmergedRSE && Remove redundant iteration in getRSEList.

In _exec check isClean status from the RSE

Add note about totalNumFiles && Removing filterRseList() method/

typo

todor-ivanov commented:
@amaltaro there is only two PRs that  need to be added in the deployment and service_config repositories so that we can put the `RSEEXPR` variable in the `config-unmerged.py` file, but this one needs to wait for merging 4 previous commits related to adding the `config-unmerged.py` itself to the repositories. So I would say you may start your review and once those are there I'll put the link in the comment at the proper section.

amaltaro commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -57,27 +70,144 @@ def execute(self):
+
+        return rse
+
+    def _checkClean(self, rse):
```
What are your plans if an RSE is marked as not cleaned? My point is, perhaps we could have a different evaluation here to avoid trying to delete 10k files when only a handful of them were missed.

todor-ivanov replied:
Actually I plan to do exactly nothing, since  we are going to stumble on this RSE on a next cycle only once the results of the current iteration is reflected everywhere down the chain. So the next time we'll have to deal with this same RSE we won't iterate through the already deleted files. 

It may happen that we delete all the files put for deletion for the current cycle, but in case the RSE was  too polluted and number of files have exceded the limit of deletions per RSE fora  a single cycle the ... well... this is definitely an RSE we may not consider clean and will have to iterate through it again start digesting what ever was an excess from the previous run. 

Or an equivalent scenario - we have some broken files to the level no deletion can happen even though they are present in the namespace. Then ... whell this is something that would need a manual intervention and untill that   happens the RSE shouldn't be considered clean either. And we need to know somehow about this case. I've not yet figured it out how to identify and signal it properly though. Any suggestions are welcome.

amaltaro replied:
My suggestion is to actually define a fraction where an RSE is considered clean. Like, if >75% (or 90%?) of the deletable files managed to be deleted, then we should consider that RSE clean, otherwise we will simply add extra load on our microservice and on the RSE storage.

I fear that the microservice will spend a long time trying to delete some final problematic files, and this is something that we should avoid IMO.

To be explicit, what I'm trying to avoid here is to iterate over a list of 100k files, in a given RSE, when you only need to delete 500 of them. I would say, in such cases, it would be better to wait for a new dump of the consistency check tool.

Just some thoughts. The True/False might not be enough for us.

todor-ivanov replied:
thanks @amaltaro 

> To be explicit, what I'm trying to avoid here is to iterate over a list of 100k files, in a given RSE, when you only need to delete 500 of them. 

As I mentioned in my previous reply, this effect should be mitigated by the `limitFilesPerRSE` configuration parameter. 

> I would say, in such cases, it would be better to wait for a new dump of the consistency check tool.

And again my point exactly - when I said:  

> we are going to stumble on this RSE on a next cycle only once the results of the current iteration is reflected everywhere down the chain

So we  are on the same page. Only that all these changes I have foreseen in the next issue, where the actual per RSE load and work would be estimated and executed, At that moment we will have a better view the situation and how to organize those limits. 

And, yes we can think of a non boolean resolution of the question what actions to be taken towards a partially cleaned RSEs.

todor-ivanov replied:
BTW we should make a note and link the next issue https://github.com/dmwm/WMCore/issues/10410 to this very comment so we  do not forget this.

amaltaro replied:
> As I mentioned in my previous reply, this effect should be mitigated by the limitFilesPerRSE configuration parameter.

The only way I can think of, to mitigate it, would be if we kept the list of files that we have successfully deleted persisted somewhere. Otherwise we have no idea which of the 100k files (from the example above) have been already deleted or not.
Maybe something to be considered when we attach this service to a database.

All right, let's keep discussing it in your next pull request then, once you implement the real file deletion logic.

Thanks for this follow up, Todor.

amaltaro commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -57,27 +70,144 @@ def execute(self):
+                continue
+        return totalNumRses, totalNumFiles, numRsesCleaned, numFilesDeleted
+
+    def filterRseList(self, rseList):
```
Doesn't the method `getRSEList` does all the job already? If not, how are you considering to use this one?

todor-ivanov replied:
I was thinking about applying the the rest of the load constraints constraints here, those like `self.msConfig.limitRSEsPerInstance` etc. We can skip it if we want, or preserve  the placeholder for whatever constrain comes next.

amaltaro replied:
I think I would rely on the expression only. If a POD is defined to work in a set of RSEs, we better make sure that very same set is cleaned up by the same POD.
Regarding placeholders for a _possible_ development, you know what my opinion is on that. Just ditch it, please ;)

todor-ivanov replied:
removing it :)

todor-ivanov replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py:
```diff
@@ -0,0 +1,49 @@
+            "name": rseName,
+            "delInterface": "",
+            "isClean": False,
+            "counters": {"totalNumFiles" : 0,
```
I like this schema!
Just one question on the counters/totalNumFiles. Is it meant to be total number of files eligible for deletion? Or total number of files in the unmerged area (meaning, not necessarily eligible for deletion)?

todor-ivanov replied:
Good question. This is a sign that the name chosen was not expressive enough, so it does not contain the needed amount of information in it. I should be total files eligible for deletion. It is used in the `_checkClean` method it is also returned to the report document (the string there tells it) ... but ye the name is maybe not the best. - I was following the general login in the report document and didn't think that would lead to the current confusion.

amaltaro replied:
As we discussed offline, this is meant to be the total amount of deletable files (files that pass the WMStats filter). I gave you a suggestion this afternoon, but you could add a comment above and it should be enough as well.

todor-ivanov replied:
note made

amaltaro reviewed (commented):
These changes are looking good, Todor. I have left some suggestions/questions in the code for your consideration though.

todor-ivanov commented on src/python/WMCore/MicroService/MSManager.py:
```diff
@@ -50,11 +50,11 @@ def daemon(func, reqStatus, interval, logger):
     while True:
         try:
-            func()
+            func(*args, **kwargs)
```
Not exactly needed, but this is the general function signature how it should look like. And this is something I regret not doing it in the commit introducing that function, so I decided to push it now.

todor-ivanov replied:
sorry we are talking about the function call here ... but still the same the generalization statement applies equally strong to both  definition and call.

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py:
```diff
@@ -0,0 +1,49 @@
+        """
+        Document Schema:
+        {
+        "name": "RSEName",
```
If you think the doc schema is too much I would just remove the whole docstring. It serves no purpose for a standard `__init__` method.

todor-ivanov replied:
done

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -57,27 +70,144 @@ def execute(self):
-        Queries CRIC for the proper RSE lists to iterate through.
-        return: List of RSEs.
+        Queries Rucio for the proper RSE list to iterate through.
+        :return: List of MSUnmergedRSE objects.
```
Which is exactly correct. 
This was  a docstring reflecting the return from this method  in a previous version when I was building the list of `MSUnmergedRSE` objects directly here. but I left that behind - because it means that   once we  start accumulating  all those huge lists of deletions and information in the objects we will have to keep them until the exit. I'd rather prefer to generate a simple list of identifiers (RSE names) here and pass it up and generate those objects right before I through them in the pipeline - this way, once the `MSUnmergedRSE` object goes through, the only leftover should be the counters.... 

Thanks for catching this docstring inconsistency. I will fix it.

todor-ivanov replied:
Fixed

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -57,27 +70,144 @@ def execute(self):
         rseList = []
-
+        try:
+            for rse in self.rucio.evaluateRSEExpression(self.msConfig['rseExpression']):
```
that is correct too

todor-ivanov replied:
The explanation why I was iterating through this list like that  was given above. I may fix it even though for the sake of staying  on the safe side the fix should include a `deep copy`.. which is going to do exactly the same as this cycle is doing.

todor-ivanov replied:
Done. Works just fine without `deepcopy` too.

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -57,27 +70,144 @@ def execute(self):
+        A simple function for zeroing the deleted files counters.
+        """
+
+        for rse in self.rseCounters:
```
Not exactly... this function was intended to not only clears the counters, but it should also initializes them if not existent. As you can see the very top in the `__init__` those counters are empty dictionaries. But the place this method is called is not the best we do not have the `rseList` yet ... etc. We need to have a quick discussion offline .. on the strategy how to use those, because they are actually duplicates of the already existing counters which are passed between function calls, but I like keeping them in the object more BTW.

amaltaro replied:
Oh, I see. You're correct!

todor-ivanov commented:
Thanks  for the review @amaltaro! I just fixed the things that  were an immediate change. but for some of the rest I may need your input, because there are several paths that   may be followed there. Please read my replies to the comments inline. Thanks again.

amaltaro commented:
@todor-ivanov can you please clean the pep8 warning reported above?

todor-ivanov commented:
Thanks for your replies @amaltaro I implemented the changes you requested. Could you please take another look.

amaltaro reviewed (approved):
Thanks for providing further changes, Todor. Everything looks good to me, please squash the commits (and amend the commit message if necessary).
I made a note on all the 3 deployment change pull requests as well, where a typo needs to be fixed. Thanks!

todor-ivanov commented:
Thanks @amaltaro
I just squashed them. And also fixed the misalignment of variable names in `msConfig['rseExpression']` here vs. `data.rseExpr` in the `config-unmerged.py` file.
And here are all the commits fixing the typo in the config files too:

https://github.com/dmwm/deployment/pull/1061/commits/03d85bf7b96e6c707ae01d95d49aa81c84cc3cdf

https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/90/diffs?commit_id=f3fd8f0fdb63269122dc8649951c71a89d8a2395

https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/91/diffs?commit_id=d26dcff6c77f13debdfde208763519795a129ab2

amaltaro commented:
Thank you very much, Todor.
